search/includes/classes/class-concurrency-limiter.php

```diff
@@ -16,27 +22,52 @@ class Concurrency_Limiter {
 	/** @var bool */
 	private $doing_request = false;
 
+	/** @var bool */
+	private $should_fail = false;
+
+	/** @var BackendInterface|null */
+	private $backend = null;
+
 	public function __construct() {
-		if ( function_exists( 'apcu_enabled' ) && apcu_enabled() ) {
-			$this->init();
-		}
+		$this->init();
 	}
 
-	public function __destruct() {
-		if ( $this->doing_request ) {
-			$this->dec_key();
+	public function init(): bool {
+		if ( defined( 'WP_TESTS_DOMAIN' ) ) {
+			$backend_class = Object_Cache_Backend::class;
+		} else {
+			$backend_class = 'Windows' === PHP_OS_FAMILY ? APCu_Backend::class : Semaphore_Backend::class;
+		}
```

Comment:
For tests, we use the most harmless class.

For Windows, we have to use APCu because PHP does not support semaphores there.

For other platforms, we use semaphores because they are the most reliable. However, once a semaphore I created, it is impossible to change its limit (and consequently, the number of concurrent connections allowed). This should not be much of an issue because all changes to the code trigger redeployment, and this will recreate a semaphore with the correct limits.